builtin/logical/transit/path_sign_verify.go

```diff
@@ -366,10 +366,10 @@ func (b *backend) pathSignWrite(ctx context.Context, req *logical.Request, d *fr
 	}
 	if !b.System().CachingDisabled() {
 		p.Lock(false)
+		defer p.Unlock()
```

Comment:
Aha! Yes, play around with the tests in my branch here: https://github.com/openbao/openbao/pull/136#issuecomment-1937357440

If you rebase them (`git rebase -i main`), you can edit them to be like so:

```
pick e37f70887 Defer policy Unlock call when Lock is called  ## < your commit
pick e97f53909 Add test cases for Transit sign/verify deadlock ## < test cases
break ## < break to let you run tests
pick 6769b473e Defer cleanup in keysutil.LockManager.GetPolicy() ## < unrelated cleanup in keysutil
pick 82fb4d715 Simplify unlock behavior in Transit with defers ## < moving all `p.Unlock()` to the consistent format
```

You can then run tests like so:

```bash
$ reset ; cd builtin/logical/transit && go test -count=1 -timeout=25s -run=TestTransit_NoDeadlock_SignVerify .
```

and see a failure if you search for this:

```
panic: test timed out after 25s
running tests:
	TestTransit_NoDeadlock_SignVerify/rsa-2048/direct (25s)
	TestTransit_NoDeadlock_SignVerify/rsa-3072/direct (25s)
	TestTransit_NoDeadlock_SignVerify/rsa-4096/direct (25s)
```

This shows that with caching disabled (not taking this if branch), we need to call unlock as well. :-)  One set of tests run using a cluster with caching enabled (default), the other runs with caching disabled. 

(you can then `--continue` the rebase to exit)

Reply:
Feel free to grab what you want from that branch and add it to this PR :-) If you want to do the move to `defer`s yourself, happy to let you do that. But if you get a moment, perhaps expanding the deadlock cases to cover the other endpoints (verify, decrypt, encrypt, hmac, ...) might be good just to make sure we didn't miss anything... Now that I remember how to test it in short time. :-D

Reply:
Yes, thanks for the explanation, in that case it completely makes sense to have the `defer`s outside of the `if` blocks.
As most of the work has already been done in your branch I'm ok with closing this one. As for the testing you are suggesting and missing `defer` in some Unlock calls I can commit those into your branch (if that's possible).